sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ShouldImplementExportedInterfaces.cs

```diff
@@ -0,0 +1,154 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ShouldImplementExportedInterfaces : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4159";
+        private const string MessageFormat = "{0} '{1}' on '{2}' or remove this export attribute.";
+        private const string ActionForInterface = "Implement";
+        private const string ActionForClass = "Derive from";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> ExportAttributes = new HashSet<KnownType>
+        {
+            KnownType.System_ComponentModel_Composition_ExportAttribute,
+            KnownType.System_ComponentModel_Composition_InheritedExportAttribute,
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var attributeSyntax = (AttributeSyntax)c.Node;
+
+                    var attributeCtorSymbol = c.SemanticModel.GetSymbolInfo(attributeSyntax.Name).Symbol as IMethodSymbol;
+                    if (attributeCtorSymbol == null ||
+                        !attributeCtorSymbol.ContainingType.IsAny(ExportAttributes))
+                    {
+                        return;
+                    }
+
+                    var exportedType = GetExportedTypeSyntax(attributeSyntax, c.SemanticModel);
+                    var attributeTargetType = GetAttributeTargetSymbol(attributeSyntax, c.SemanticModel);
+
+                    if (exportedType != null &&
+                        attributeTargetType != null &&
+                        !attributeTargetType.DerivesOrImplements(exportedType))
+                    {
+                        var action = exportedType.IsInterface()
+                            ? ActionForInterface
+                            : ActionForClass;
+
+                        c.ReportDiagnosticWhenActive(
+                            Diagnostic.Create(rule, attributeSyntax.GetLocation(), action, exportedType.Name, attributeTargetType.Name));

```

Comment:
[cosmetic] Why this line split?

Reply:
@Evangelink because it is too long otherwise

Reply:
It is long like this too, though 💃